apps/web/pages/auth/setup.tsx

```diff
@@ -32,11 +32,22 @@ const SetupFormStep1 = (props: { setIsLoading: (val: boolean) => void }) => {
   const { t } = useLocale();
 
   const formSchema = z.object({
-    username: z.string().min(1, t("at_least_characters", { count: 1 })),
+    username: z
+      .string()
+      .refine((val) => val.trim().length >= 1, { message: t("at_least_characters", { count: 1 }) }),
     email_address: z.string().email({ message: t("enter_valid_email") }),
     full_name: z.string().min(3, t("at_least_characters", { count: 3 })),
-    password: z.string().refine((val) => isPasswordValid(val.trim()), {
-      message: t("invalid_password_hint"),
+    password: z.string().superRefine((data, ctx) => {
+      const result = isPasswordValid(data, true);
+      Object.keys(result).map((key: string) => {
+        if (!result[key as keyof typeof result]) {
+          ctx.addIssue({
+            code: z.ZodIssueCode.custom,
+            path: [key],
+            message: t(`password_hint_${key}`),
```

Comment:
Basically constructing custom errors when checks fail from the password validation function, now that the function has the ability to return those checks.